src/text-input/index.ts

```diff
@@ -231,34 +272,60 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
 		});
 	}
 
+	protected renderHelperText(): DNode {
+		const { properties: { helperText }, _state: { message } } = this;
+		const text = message || helperText;
+		const invalid = !!message;
```

Comment:
why don't you use the `_state.valid` boolean here?

Reply:
Fixed.